src/thx/Nel.hx

```diff
@@ -44,23 +61,100 @@ abstract Nel<A> (NonEmptyList<A>) from NonEmptyList<A> to NonEmptyList<A> {
       case ConsNel(x, xs): f(x, xs.fold(f));
     };
 
-  @:op(N+N0) 
-  public function append(nel: NonEmptyList<A>): Nel<A> 
+  @:op(N+N0)
+  public function append(nel: NonEmptyList<A>): Nel<A>
     // TODO: Implement imperatively for stack safety.
     return switch this {
       case Single(x): ConsNel(x, nel);
       case ConsNel(x, xs): ConsNel(x, xs.append(nel));
     };
 
+/**
+Gets the head item of the non-empty list, which is guaranteed to exist
+**/
+  public function head() : A {
+    return switch this {
+      case Single(x) : x;
+      case ConsNel(x, xs) : x;
+    };
+  }
+
+/**
+Gets the tail (all but the first element) of the `Nel<A>` as an `Option<Nel<A>>`
+**/
+  public function tail() : Option<Nel<A>> {
+    return switch this {
+      case Single(x) : None;
+      case ConsNel(x, xs) : Some(xs);
+    };
+  }
+
+/**
+Gets the initial elements (all but the last element) of the `Nel<A>` as an `Option<Nel<A>>`
+**/
+  public function init() : Option<Nel<A>> {
+    return switch this {
+      case Single(x) : None;
+      case ConsNel(x, xs) : switch (xs : Nel<A>).init() {
+        case Some(xsInit) : Some(ConsNel(x, xsInit));
+        case None : Some(Single(x));
+      };
+    };
+  }
+
+/**
+Gets the last item of the `Nel<A>`, which is guaranteed to exist.
+**/
+  public function last() : A {
+    return switch this {
+      case Single(x) : x;
+      case ConsNel(x, xs) : xs.last();
+    };
+  }
+
+/**
+Returns a new `Nel<A>` with the given item added at the end.
+Does not modify this `Nel<A>`.
+**/
+  public function push(a : A) : Nel<A> {
+    return append(Single(a));
```

Comment:
Since append is O(N), it's probably necessary to add a warning that this shouldn't be used most of the time. It would be better to add a function to allow O(1) cons.